### PR TITLE
Feat/expose pure esm and ipfs upgrade

### PR DESCRIPTION
**Pull request by stephhuynh18**

### CODE REVIEW

**Comment by ukstv** [comment]
It is huge.

**Comment by oed** [comment]
Can we have a list of tests that are disabled?

**Comment by ukstv** [comment]
These ones:
- `packages/blockchain-utils-validation/src/blockchains/__tests__/tezos.test.ts`,
- `packages/canary-integration/src/__tests__/caip10/tezos.test.ts`,
- `packages/canary-integration/src/__tests__/caip10/ethereum.test.ts`,
- `packages/blockchain-utils-validation/src/blockchains/__tests__/ethereum.test.ts`

Added the list to Ceramic 2.0 release Notion page.

**Comment by oed** [comment]
hm, the ethereum ones are worriesome.

**Comment by ukstv** [comment]
The common theme is they all use mocks, and Jest can barely mock an ES module.

File: packages/cli/src/ceramic-daemon.ts [hunk 1]
```diff
@@ -1,33 +1,35 @@
...
-import PkhDidResolver from 'pkh-did-resolver'
+import { StreamID, StreamType } from '@ceramicnetwork/streamid'
+import * as ThreeIdResolver from '@ceramicnetwork/3id-did-resolver'
+import * as KeyDidResolver from 'key-did-resolver'
+import * as PkhDidResolver from 'pkh-did-resolver'
 import EthrDidResolver from 'ethr-did-resolver'
-import NftDidResolver from 'nft-did-resolver'
-import SafeDidResolver from 'safe-did-resolver'
+import * as NftDidResolver from 'nft-did-resolver'
+// import SafeDidResolver from 'safe-did-resolver'
```
Comment (oed) [inline packages/cli/src/ceramic-daemon.ts:21]: Somethings up with safe DID?

File: packages/cli/src/ceramic-daemon.ts [hunk 2]
```diff
@@ -207,6 +209,8 @@ export class CeramicDaemon {
       opts.ipfs?.host
     )
 
+    console.log('ipfs', ipfs)
```
Comment (oed) [inline packages/cli/src/ceramic-daemon.ts:212]: Remove?

File: packages/core/src/__tests__/caip10-link.test.ts [hunk 3]
```diff
@@ -58,35 +55,24 @@ describe('Ceramic API', () => {
...
-      expect(link.metadata.controllers).toHaveLength(1)
-      expect(link.metadata.controllers[0]).toEqual(
-        '0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1'
-      )
-      expect(link.did).toBeNull()
-      expect(link.state.log).toHaveLength(1)
-      expect(link.state).toMatchSnapshot()
-    })
-
-    it('Create from legacy account id', async () => {
```
Comment (oed) [inline packages/core/src/__tests__/caip10-link.test.ts:83]: Why was this test removed?

File: packages/core/src/__tests__/caip10-link.test.ts [hunk 4]
```diff
@@ -108,94 +94,67 @@ describe('Ceramic API', () => {
...
-      const linkProof = { account, did: ceramic.did.id }
-      authProvider.createLink.mockReturnValueOnce(linkProof)
-
-      const link = await Caip10Link.fromAccount(ceramic, account)
-      await link.setDid(ceramic.did, authProvider, { anchor: false })
-
-      expect(link.did).toEqual(ceramic.did.id)
-      expect(link.state.log).toHaveLength(2)
-      expect(authProvider.createLink).toHaveBeenCalledTimes(1)
-      expect(validateLink).toHaveBeenCalledTimes(1)
```
Comment (oed) [inline packages/core/src/__tests__/caip10-link.test.ts:138]: Why removed?

**Review by oed — COMMENTED** [review]
looks good, minor comments

File: packages/cli/src/ceramic-daemon.ts (hunk 1 above)
Comment (ukstv) [reply packages/cli/src/ceramic-daemon.ts]: It was not released. Thanks for catching that. Restored it back.

File: packages/cli/src/ceramic-daemon.ts (hunk 2 above)
Comment (ukstv) [reply packages/cli/src/ceramic-daemon.ts]: nice. thanks. removed.

File: packages/core/src/__tests__/caip10-link.test.ts (hunk 4 above)
Comment (ukstv) [reply packages/core/src/__tests__/caip10-link.test.ts:138]: I think, it is a remnant of merging `develop` into 2.x release branch at some point. I remember catching version change, but this one I did not see.

Comment (oed) [reply packages/core/src/__tests__/caip10-link.test.ts:138]: seems important to make sure we keep support for the new caip10 format. Can we add the test back?

Comment (ukstv) [reply packages/core/src/__tests__/caip10-link.test.ts:138]: But, no, this theory does not stand. I do not know.

Comment (ukstv) [reply packages/core/src/__tests__/caip10-link.test.ts:138]: Yup, adding it back.

Comment (ukstv) [reply packages/core/src/__tests__/caip10-link.test.ts]: Recovered.

File: packages/blockchain-utils-validation/jest.config.json [hunk 5]
```diff
@@ -1,9 +1,12 @@
 {
-  "moduleFileExtensions": ["js", "json", "ts"],
-  "testRegex": ".(spec|test).ts$",
   "testEnvironment": "node",
+  "resolver": "jest-resolver-enhanced",
+  "testMatch": ["**/?(*.)+(spec|test).[jt]s?(x)"],
+  "extensionsToTreatAsEsm": [".ts"],
   "moduleNameMapper": {
-    "^@zondax/filecoin-signing-tools/nodejs$": "@zondax/filecoin-signing-tools"
+    "^(\\.{1,2}/.*)\\.js$": "$1"
```
Comment (stbrody) [inline packages/blockchain-utils-validation/jest.config.json:7]: man I hate regexes.  What does this do?  I wish you could add a comment explaining this, but I'm guessing comments aren't allowed here because JSON :sob:

File: packages/blockchain-utils-validation/src/blockchains/__tests__/ethereum.test.ts [hunk 6]
```diff
@@ -54,21 +54,22 @@ const send = (provider: any, data: any): Promise<any> =>
...
-        return new originalModule.Web3Provider(lazyProvider())
-      },
-    }
-  }
-  return {
-    ...originalModule,
-    getNetwork,
-  }
-})
+// TODO ESM Mocking
```
Comment (stbrody) [inline packages/blockchain-utils-validation/src/blockchains/__tests__/ethereum.test.ts:57]: can we put a ticket # in the TODO here?  Can be one ticket for tracking all the little odds and ends and todos that are left to get this ready to go.  That way when we can grep for `TODO(<ticket number>)` and see nothing show up we know we've gotten everything

Comment (stbrody) [reply packages/blockchain-utils-validation/src/blockchains/__tests__/ethereum.test.ts:57]: ditto for all other TODOs in this file (and elsewhere in this PR)

File: packages/blockchain-utils-validation/src/index.ts [hunk 7]
```diff
@@ -1,11 +1,11 @@
 import { LinkProof } from '@ceramicnetwork/blockchain-utils-linking'
-import ethereum from './blockchains/ethereum'
-import filecoin from './blockchains/filecoin'
-import polkadot from './blockchains/polkadot'
-import eosio from './blockchains/eosio'
-import cosmos from './blockchains/cosmos'
-import near from './blockchains/near'
-import tezos from './blockchains/tezos'
+import { handler as ethereum } from './blockchains/ethereum.js'
```
Comment (stbrody) [inline packages/blockchain-utils-validation/src/index.ts:2]: optional nitpick: any way we could give these a better default name?

File: packages/canary-integration/src/__tests__/caip10/ethereum.test.ts [hunk 8]
```diff
@@ -54,21 +54,22 @@ const send = (provider: any, data: any): Promise<any> =>
...
-        return new originalModule.Web3Provider(lazyProvider())
-      },
-    }
-  }
-  return {
-    ...originalModule,
-    getNetwork,
-  }
-})
+// TODO ESM Mocking
```
Comment (stbrody) [inline packages/canary-integration/src/__tests__/caip10/ethereum.test.ts:57]: ticket #

File: packages/canary-integration/src/__tests__/caip10/tezos.test.ts [hunk 9]
```diff
@@ -1,58 +1,60 @@
...
-test('wrong proof', async () => {
-  const authProvider = new TezosAuthProvider(provider)
-  await wrongProof(ceramic, authProvider)
-}, 120000)
-
-test('clear did', async () => {
-  const authProvider = new TezosAuthProvider(provider)
-  await clearDid(ceramic, authProvider)
-}, 120000)
+// TODO ESM Mocking
```
Comment (stbrody) [inline packages/canary-integration/src/__tests__/caip10/tezos.test.ts:1]: ticket #

File: packages/cli/bin/ceramic.js [hunk 10]
```diff
@@ -1,3 +1,3 @@
 #!/usr/bin/env node
 
-require('../lib/bin/ceramic')
+await import('../lib/bin/ceramic.js')
```
Comment (stbrody) [inline packages/cli/bin/ceramic.js:3]: woah.  What does `await import` do and when is it necessary?  Why do we use this here but no where else in this PR?

File: packages/cli/package.json [hunk 11]
```diff
@@ -42,49 +43,44 @@
...
     "reflect-metadata": "^0.1.13",
     "s3leveldown": "^2.2.2",
-    "safe-did-resolver": "^0.0.5",
+    "safe-did-resolver": "^1.0.0-alpha.0",
     "stream-to-array": "^2.3.0",
     "typedjson": "^1.8.0",
     "uint8arrays": "^3.0.0"
   },
   "devDependencies": {
+    "@ceramicnetwork/canary-integration": "^2.0.0-alpha.0",
```
Comment (stbrody) [inline packages/cli/package.json:74]: this seems backwards.  I thought the idea with canary-integration was that it could have a dependency on all our other packages and be a unified place for integration tests.  Putting this dependency here would prevent us from ever being able to import the `cli` package into a canary integration test without introducing a circular dependency.  Why do we need to add this?

File: packages/cli/src/__tests__/ceramic-daemon.test.ts [hunk 12]
```diff
@@ -57,7 +58,7 @@ describe('Ceramic interop: core <> http-client', () => {
 
   beforeAll(async () => {
     tmpFolder = await tmp.dir({ unsafeCleanup: true })
```
Comment (stbrody) [inline packages/cli/src/__tests__/ceramic-daemon.test.ts:60]: unused?

File: packages/cli/src/__tests__/ceramic-daemon.test.ts [hunk 13]
```diff
@@ -278,7 +279,7 @@ describe('Ceramic interop: core <> http-client', () => {
     )
 
     // Load v1 (anchor on top of genesis commit)
-    const v1Id = doc.id.atCommit(doc.state.log[1].cid)
+    const v1Id = new CommitID(doc.id.type, doc.id.cid, doc.state.log[1].cid)
```
Comment (stbrody) [inline packages/cli/src/__tests__/ceramic-daemon.test.ts:282]:
```suggestion
    const v1Id = CommitID.make(doc.id, doc.state.log[1].cid)
```

Comment (stbrody) [reply packages/cli/src/__tests__/ceramic-daemon.test.ts:282]: here and below throughout this file.  Seems more readable to me.

File: packages/cli/src/ceramic-daemon.ts [hunk 14]
```diff
@@ -323,12 +325,17 @@ export class CeramicDaemon {
...
-    res.json({ streamId: stream.id.toString(), state: StreamUtils.serializeState(stream.state) })
+    try {
+      const stream = await this.ceramic.createStreamFromGenesis(
+        type,
+        StreamUtils.deserializeCommit(genesis),
+        opts
+      )
+      res.json({ streamId: stream.id.toString(), state: StreamUtils.serializeState(stream.state) })
+    } catch (e) {
+      console.error(e)
```
Comment (stbrody) [inline packages/cli/src/ceramic-daemon.ts:336]: This feels like a weird change to come in this PR, and only to this one function but not all the other similar functions in this file.  Seems like we should have a separate PR that adds error logging to all HTTP endpoints if it's not already there?  Although TBH I thought we already did log errors as part of the http handler

File: packages/core/src/__tests__/ceramic-api.test.ts [hunk 15]
```diff
@@ -392,10 +402,6 @@ describe('Ceramic API', () => {
           streamId: streamA.id,
           paths: ['/b/d', '/notExistStreamId'],
         },
-        {
-          streamId: notExistStreamId,
```
Comment (stbrody) [inline packages/core/src/__tests__/ceramic-api.test.ts:396]: why was this removed?

File: packages/core/src/dispatcher.ts [hunk 16]
```diff
@@ -171,6 +182,7 @@ export class Dispatcher {
       try {
         dagResult = await this._ipfs.dag.get(asCid, { timeout: IPFS_GET_TIMEOUT, path })
       } catch (err) {
+        console.log('_getFromIpfs:error', err)
```
Comment (stbrody) [inline packages/core/src/dispatcher.ts:185]: rm

File: packages/core/src/dispatcher.ts [hunk 17]
```diff
@@ -311,6 +323,7 @@ export class Dispatcher {
    */
   async close(): Promise<void> {
     this.messageBus.unsubscribe()
+    await this.tasks.onIdle()
```
Comment (stbrody) [inline packages/core/src/dispatcher.ts:326]: oooh, does this fix https://github.com/ceramicnetwork/js-ceramic/issues/1963?

File: packages/core/src/pubsub/incoming-channel.ts [hunk 18]
```diff
@@ -48,7 +48,9 @@ export class PubsubIncoming extends Observable<IPFSPubsubMessage> {
       return () => {
         this.tasks.clear()
         this.tasks.add(async () => {
-          await ipfs.pubsub?.unsubscribe(topic, onMessage)
+          await ipfs.pubsub?.unsubscribe(topic, onMessage).catch(() => {
+            // Do Nothing
```
Comment (stbrody) [inline packages/core/src/pubsub/incoming-channel.ts:52]: should we at least log a warning if something goes wrong with the unsubscribe?

File: packages/core/src/pubsub/pubsub.ts [hunk 19]
```diff
@@ -54,10 +55,11 @@ export class Pubsub extends Observable<PubsubMessage> {
     private readonly topic: string,
     private readonly resubscribeEvery: number,
     private readonly pubsubLogger: ServiceLogger,
-    private readonly logger: DiagnosticsLogger
+    private readonly logger: DiagnosticsLogger,
+    readonly tasks: TaskQueue = new TaskQueue()
   ) {
     super((subscriber) => {
-      const incoming$ = new IncomingChannel(ipfs, topic, resubscribeEvery, pubsubLogger, logger)
+      const incoming$ = new IncomingChannel(ipfs, topic, resubscribeEvery, pubsubLogger, logger, tasks)
```
Comment (stbrody) [inline packages/core/src/pubsub/pubsub.ts:62]: now that we seem to be passing in a shared task queue from higher levels down into each lower level, should we remove the default value for the `tasks` arg on `IncomingChannel` and similar to force callers to pass the queue explicitly?

Comment (stbrody) [reply packages/core/src/pubsub/pubsub.ts:62]: (note: doing that in this PR is totally optional, it's not a priority just wondering about a potential tightening of the API to make it harder to accidentally misuse)

File: packages/core/src/store/level-state-store.ts [hunk 20]
```diff
@@ -1,17 +1,20 @@
...
+import type Level from 'level-ts'
 import { StreamState, StreamStateHolder, StreamUtils } from '@ceramicnetwork/common'
-import { StateStore } from './state-store'
-import StreamID from '@ceramicnetwork/streamid'
+import { StateStore } from './state-store.js'
+import { StreamID } from '@ceramicnetwork/streamid'
 import * as fs from 'fs'
 import path from 'path'
 
+const LevelC = (levelTs as any).default as unknown as typeof Level
```
Comment (stbrody) [inline packages/core/src/store/level-state-store.ts:9]: what does `LevelC` mean?  Can we give this a more descriptive name?

File: packages/core/src/store/level-state-store.ts [hunk 21]
```diff
@@ -1,17 +1,20 @@
...
+const defaultLevelFactory = (path: string) => new LevelC(path)
 /**
  * Ceramic store for saving stream state to a local leveldb instance
  */
 export class LevelStateStore implements StateStore {
   #store: Level
 
-  constructor(private storeRoot: string) {}
+  constructor(private storeRoot: string, private readonly levelFactory = defaultLevelFactory) {}
```
Comment (stbrody) [inline packages/core/src/store/level-state-store.ts:17]: why do we need this extra level of indirection of a level factory?

**Review by stbrody — CHANGES_REQUESTED** [review]
mostly looks good, a few small comments throughout.  Probably my biggest concern is adding the dep on canary-integration to the cli package.

File: packages/canary-integration/src/__tests__/caip10/ethereum.test.ts (hunk 8 above)
Comment (ukstv) [reply packages/canary-integration/src/__tests__/caip10/ethereum.test.ts]: done.

File: packages/blockchain-utils-validation/src/index.ts (hunk 7 above)
Comment (ukstv) [reply packages/blockchain-utils-validation/src/index.ts:2]: We call this thing a handler, and the current structure has proven to be easy enough for developers to grok and extend, so please, let it remain it as it is, unless you have some better ideas.

File: packages/core/src/dispatcher.ts (hunk 17 above)
Comment (ukstv) [reply packages/core/src/dispatcher.ts:326]: yes.

File: packages/core/src/store/level-state-store.ts (hunk 20 above)
Comment (ukstv) [reply packages/core/src/store/level-state-store.ts:19]: Added comment there:

> When Node.js imports a CJS module from ESM, it considers whole contents of `module.exports` as ESM default export.
> 'level-ts' is a CommomJS module, which exports Level constructor as `exports.default`.
> This `default` name has no special meaning from ESM perspective. It is just yet another name.
> Types provided by level-ts though make it appear as ESM default.
> 
> So, here we untangle this mess, even if ugly.
> We import type information separately from code information, and then make sure we can access
> a properly typed constructor of `Level` (thus `LevelC`) exported from level-ts package.
>
> See also https://github.com/nodejs/node/blob/master/doc/api/esm.md#commonjs-namespaces

File: packages/core/src/store/level-state-store.ts (hunk 21 above)
Comment (ukstv) [reply packages/core/src/store/level-state-store.ts:31]: Added comment there:

> Jest can barely mock modules when running in ESM mode. Some of our tests though rely on `level-ts` package
> being mocked and spied on. Instead of making testing code incredibly ugly, we make dependency on `level-ts` explicit.
> Now we can pass a carefully crafted mock object to spy on.

To be honest, this is a good practice anyway to make such external dependencies somewhat isolated.

File: packages/cli/bin/ceramic.js (hunk 10 above)
Comment (ukstv) [reply packages/cli/bin/ceramic.js:3]: You can not `require` an ESM module, that is why `import`. 

You could do `import * as something from '../path.js'`, but the next person reading this would ask what `something`, is and why it is needed here. You could disregard `await` and just do `import`, this would do the job mostly. When something wrong happens though, `await import` reports a path to this particular js file, which is useful in some situations.

File: packages/cli/package.json (hunk 11 above)
Comment (ukstv) [reply packages/cli/package.json]: removed. not needed in any way.

File: packages/cli/src/__tests__/ceramic-daemon.test.ts (hunk 12 above)
Comment (ukstv) [reply packages/cli/src/__tests__/ceramic-daemon.test.ts:60]: line 70: `core = await makeCeramicCore(ipfs, tmpFolder.path)`

File: packages/core/src/__tests__/ceramic-api.test.ts (hunk 15 above)
Comment (ukstv) [reply packages/core/src/__tests__/ceramic-api.test.ts]: Good catch, thanks. With go-ipfs somehow timeout is not respected and the test hangs. Worrying. Created an issue to track the fix.

File: packages/cli/src/ceramic-daemon.ts (hunk 14 above)
Comment (ukstv) [reply packages/cli/src/ceramic-daemon.ts]: "fog of war" remnant. restored.

File: packages/core/src/store/level-state-store.ts (hunk 20 above)
Comment (stbrody) [reply packages/core/src/store/level-state-store.ts:19]: Ah got it, that's helpful thanks!  Can we call it LevelConstructor then?  LevelC doesn't really convey any meaning to me

Comment (stbrody) [reply packages/core/src/store/level-state-store.ts:19]: Oh I see, you actually added all that as a comment to the code.  That's great and addresses my concern, thanks!
